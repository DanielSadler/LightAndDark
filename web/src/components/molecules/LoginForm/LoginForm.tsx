import React, { FC, useState } from 'react';
import { useFormik } from 'formik';
import { useRouter } from 'next/router';
import { Button, TextInput, Notification, Checkbox } from '../../atoms';
import { useLoginMutation } from '../../../graphql';

type Props = {};

export const LoginForm: FC<Props> = () => {
  const [rememberMe, setRememberMe] = useState(false);
  const [loginError, setLoginError] = useState(false);
  const router = useRouter();
  const { mutate: login } = useLoginMutation({
    onSuccess: (data) => {
      if (data.login.access_token && data.login.refresh_token) {
        localStorage.setItem('access_token', data.login.access_token);
        localStorage.setItem('refresh_token', data.login.refresh_token);
        router.push('/home');
      }
    },
    onError: () => setLoginError(true),
  });
  const formik = useFormik({
    initialValues: {
      email: '',
      password: '',
    },
    onSubmit: (loginData) => login({ loginData }),
  });
  return (
    <div className="mt-8 sm:mx-auto sm:w-full sm:max-w-md">
      <div className="bg-white py-8 px-4 shadow sm:rounded-lg sm:px-10">
        <div className="space-y-6">
          <TextInput
            label="Email address"
            id="email"
            name="email"
            type="email"
            autoComplete="email"
            required
            onChange={formik.handleChange}
            value={formik.values.email}
          />
          <TextInput
            label="Password"
            id="password"
            name="password"
            type="password"
            autoComplete="current-password"
            required
            onChange={formik.handleChange}
            value={formik.values.password}
          />

          <div className="flex items-center justify-between">
            <Checkbox
              label="Remember me"
              defaultChecked={rememberMe}
              onChange={(current) =>
                setRememberMe(Boolean(current.target.value))
              }
              data-testid="remember-me-checkbox"
            />

            <div className="text-sm">
              <a
                href="#"
                className="font-medium text-indigo-600 hover:text-indigo-500"
              >
                Forgot your password?
              </a>
            </div>
          </div>

          <div>
            <Button onClick={() => formik.handleSubmit()}>Sign in</Button>
          </div>
          {loginError && (
            <Notification
              setShow={setLoginError}
              error={true}
              headerText="Incorrect username or password"
            />
          )}
        </div>
      </div>
    </div>
  );
};
