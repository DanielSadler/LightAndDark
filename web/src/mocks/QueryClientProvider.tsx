import React from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';

interface Props {
  children: React.ReactNode;
}

const client = new QueryClient();

export const Wrapper = ({ children }: Props) => (
  <QueryClientProvider client={client}>{children}</QueryClientProvider>
);
