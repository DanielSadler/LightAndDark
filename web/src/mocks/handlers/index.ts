import { exampleQuery } from './Example';

export const handlers = [exampleQuery];
