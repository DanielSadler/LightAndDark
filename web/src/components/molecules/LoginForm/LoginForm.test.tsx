import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { expect } from '@jest/globals';
import { LoginForm } from './LoginForm';
import { Wrapper } from '../../../mocks/QueryClientProvider';

describe('LoginForm', () => {
  it('should render', () => {
    render(
      <Wrapper>
        <LoginForm />
      </Wrapper>
    );
  });
  it('should match a snapshot', () => {
    const tree = renderer
      .create(
        <Wrapper>
          <LoginForm />
        </Wrapper>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
