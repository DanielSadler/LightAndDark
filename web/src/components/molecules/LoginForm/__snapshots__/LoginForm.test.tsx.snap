// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match a snapshot 1`] = `
<div
  className="mt-8 sm:mx-auto sm:w-full sm:max-w-md"
>
  <div
    className="bg-white py-8 px-4 shadow sm:rounded-lg sm:px-10"
  >
    <div
      className="space-y-6"
    >
      <div>
        <label
          className="block text-sm font-medium text-gray-700"
          htmlFor="email"
        >
          Email address
        </label>
        <div
          className="mt-1"
        >
          <input
            autoComplete="email"
            className="appearance-none block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm placeholder-gray-400 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
            id="email"
            name="email"
            onChange={[Function]}
            required={true}
            type="email"
            value=""
          />
        </div>
      </div>
      <div>
        <label
          className="block text-sm font-medium text-gray-700"
          htmlFor="email"
        >
          Password
        </label>
        <div
          className="mt-1"
        >
          <input
            autoComplete="current-password"
            className="appearance-none block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm placeholder-gray-400 focus:outline-none focus:ring-indigo-500 focus:border-indigo-500 sm:text-sm"
            id="password"
            name="password"
            onChange={[Function]}
            required={true}
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        className="flex items-center justify-between"
      >
        <div
          className="flex items-center"
        >
          <input
            className="h-4 w-4 text-indigo-600 focus:ring-indigo-500 border-gray-300 rounded"
            data-testid="remember-me-checkbox"
            defaultChecked={false}
            id="remember-me"
            name="remember-me"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="ml-2 block text-sm text-gray-900"
            data-testid="remember-me-checkbox-label"
            htmlFor="remember-me"
          >
            Remember me
          </label>
        </div>
        <div
          className="text-sm"
        >
          <a
            className="font-medium text-indigo-600 hover:text-indigo-500"
            href="#"
          >
            Forgot your password?
          </a>
        </div>
      </div>
      <div>
        <button
          className="w-full flex justify-center py-2 px-4 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-indigo-600 hover:bg-indigo-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
          onClick={[Function]}
          type="submit"
        >
          Sign in
        </button>
      </div>
    </div>
  </div>
</div>
`;
